azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -160,38 +198,142 @@ def __get_agent_family_manifests(self, goal_state):
-        return largest_version
+    def run(self, goal_state):
+
+        try:
+            # Ignore new agents if update is disabled. The latter flag only used in e2e tests.
+            if not self._autoupdate_enabled or not conf.get_download_new_agents():
+                return
+
+            agent_family = self._get_agent_family_manifest(goal_state)
+            version = self._get_version_from_gs(agent_family)
+            gs_id = goal_state.extensions_goal_state.id
+            self._update_rsm_version_state_if_changed(goal_state.extensions_goal_state.created_on_timestamp, agent_family)
+            # if version is specified and vm is enabled for rsm upgrades, use rsm update path, else sef-update
+            if version is None and self._is_vm_enabled_for_rsm_upgrades and self._is_version_from_rsm:
+                raise AgentUpdateError("VM Enabled for RSM upgrades but version is missing in Goal state: {0}, so skipping agent update".format(gs_id))
+            elif conf.get_enable_ga_versioning() and self._is_vm_enabled_for_rsm_upgrades:
+                updater = RSMVersionUpdater(gs_id, agent_family, None, version, self.update_state, self._is_version_from_rsm, self._daemon_version)
+                self.update_state.last_attempted_update_version = version
+            else:
+                updater = SelfUpdateVersionUpdater(gs_id, agent_family, None, None, self.update_state)
+
+            # verify if agent update is allowed
+            if not updater.should_update_agent(goal_state):
+                return
+            updater.log_new_agent_update_message()
+            updater.purge_extra_agents_from_disk()
+            agent = updater.download_and_get_new_agent(self._protocol, goal_state)
+            if agent.is_blacklisted or not agent.is_downloaded:
+                msg = "Downloaded agent version is in bad state : {0} , skipping agent update".format(
+                    str(agent.version))
+                raise AgentUpdateError(msg)
+            updater.proceed_with_update()
+
+        except Exception as err:
+            if isinstance(err, AgentUpgradeExitException):
+                raise err
+            elif isinstance(err, AgentUpdateError):
+                error_msg = ustr(err)
+            else:
+                error_msg = "Unable to update Agent: {0}".format(textutil.format_exception(err))
+            logger.warn(error_msg)
+            add_event(op=WALAEventOperation.AgentUpgrade, is_success=False, message=error_msg, log_event=False)
+            self.update_state.last_attempted_update_error_msg = error_msg
+
+    def get_vmagent_update_status(self):
+        """
+        This function gets the VMAgent update status as per the last attempted update.
+        Returns: None if fail to report or update never attempted with rsm version specified in GS
+        """
+        try:
+            if conf.get_enable_ga_versioning() and self._is_vm_enabled_for_rsm_upgrades and self._is_version_from_rsm:
```

Comment:
why do we need to check for self._is_version_from_rsm here? Does the protocol with CRP require that we do not report status when _is_vm_enabled_for_rsm_upgrades is true but _is_version_from_rsm is not?